### PR TITLE
bgpv1: Retry peer checks in NeighborAddDel test to avoid flakes

### LINKED ISSUE #25637: CI: Test_NeighborAddDel failure
## CI failure

Hit in https://github.com/cilium/cilium/pull/25469: https://jenkins.cilium.io/job/Cilium-PR-Runtime-kernel-net-next/123/testReport/(root)/Suite-runtime/RuntimeDatapathPrivilegedUnitTests_Run_Tests/

```
[07:37:34] ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
[07:37:34] ┃   FAIL  package: github.com/cilium/cilium/pkg/bgpv1/test   ┃
[07:37:34] ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
[07:37:34] 
[07:37:34] --- FAIL: Test_NeighborAddDel (9.14s)
[07:37:34] 
[07:37:34] time="2023-05-24T07:33:39Z" level=info msg="deleting dummy links"
[07:37:34] time="2023-05-24T07:33:39Z" level=info msg="adding dummy links"
[07:37:34] time="2023-05-24T07:33:39Z" level=info msg="GoBGP test instance: starting"
[07:37:34] time="2023-05-24T07:33:39Z" level=info msg="Add a peer configuration" Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGPsubsys=basic
[07:37:34] time="2023-05-24T07:33:39Z" level=debug msg="IdleHoldTimer expired" Duration=0 Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGPsubsys=basic
[07:37:34] time="2023-05-24T07:33:39Z" level=info msg="GoBGP test instance: starting"
[07:37:34] time="2023-05-24T07:33:39Z" level=info msg="Add a peer configuration" Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGPsubsys=basic
[07:37:34] time="2023-05-24T07:33:39Z" level=debug msg="IdleHoldTimer expired" Duration=0 Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGPsubsys=basic
[07:37:34] time="2023-05-24T07:33:39Z" level=debug msg="state changed" Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGP new=BGP_FSM_ACTIVEold=BGP_FSM_IDLE reason=idle-hold-timer-expired subsys=basic
[07:37:34] level=info msg=Invoked duration="78.81µs" function="test.newFixture.func6 (fixtures.go:155)" subsys=hive
[07:37:34] level=info msg=Starting subsys=hive
[07:37:34] level=info msg="Start hook executed" duration="1.682µs" function="*resource.resource[*github.com/cilium/cilium/pkg/k8s/slim/k8s/api/corev1.Service].Start" subsys=hive
[07:37:34] time="2023-05-24T07:33:39Z" level=debug msg="state changed" Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGP new=BGP_FSM_ACTIVEold=BGP_FSM_IDLE reason=idle-hold-timer-expired subsys=basic
[07:37:34] level=info msg="Start hook executed" duration=101.056956ms function="*manager.diffStore[*github.com/cilium/cilium/pkg/k8s/slim/k8s/apicore/v1.Service].Start" subsys=hive
[07:37:34] level=info msg="Start hook executed" duration="1.257µs" function="*resource.resource[*github.com/cilium/cilium/pkg/k8s/apis/cilium.iov2alpha1.CiliumBGPPeeringPolicy].Start" subsys=hive
[07:37:34] level=info msg="Start hook executed" duration="1.215µs" function="*resource.resource[*github.com/cilium/cilium/pkg/k8s/slim/k8s/api/corev1.Node].Start" subsys=hive
[07:37:34] level=info msg="Start hook executed" duration=427ns function="*resource.resource[*github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2CiliumNode].Start" subsys=hive
[07:37:34] level=info msg="Start hook executed" duration="38.391µs" function="*agent.kubernetesNodeSpecer.Start" subsys=hive
[07:37:34] level=info msg="Start hook executed" duration=101.291452ms function="*agent.Controller.Start" subsys=hive
[07:37:34] 
[07:37:34]   --- FAIL: Test_NeighborAddDel/add_two_neighbors (8.80s)
[07:37:34] 
[07:37:34] time="2023-05-24T07:33:39Z" level=info msg="GoBGP test instance: Peer Event: {0 IDLE}"
[07:37:34] level=info msg="Cilium BGP Control Plane Controller now running..." component=Controller.Run subsys=bgp-control-plane
[07:37:34] level=info msg="Cilium BGP Control Plane Controller woken for reconciliation" component=Controller.Run subsys=bgp-control-plane
[07:37:34] level=info msg="Registering BGP servers for policy with local ASN 65001" component=manager.registerBGPServer subsys=bgp-control-plane
[07:37:34] level=info msg="Reconciling peers for virtual router with local ASN 65001" component=manager.neighborReconciler subsys=bgp-control-plane
[07:37:34] level=info msg="Adding peer 172.16.100.2/32 65011 to local ASN 65001" component=manager.neighborReconciler subsys=bgp-control-plane
[07:37:34] level=info msg="Add a peer configuration" Key=172.16.100.2 Topic=Peer asn=65001 component=gobgp.BgpServerInstance subsys=bgp-control-plane
[07:37:34] level=info msg="Done reconciling peers for virtual router with local ASN 65001" component=manager.neighborReconcilersubsys=bgp-control-plane
[07:37:34] level=info msg="Successfully registered GoBGP servers for policy with local ASN 65001" component=manager.registerBGPServersubsys=bgp-control-plane
[07:37:34] time="2023-05-24T07:33:39Z" level=info msg="GoBGP test instance: Peer Event: {65001 IDLE}"
[07:37:34] time="2023-05-24T07:33:39Z" level=info msg="GoBGP test instance: Peer Event: {65001 ACTIVE}"
[07:37:34] level=info msg="Cilium BGP Control Plane Controller woken for reconciliation" component=Controller.Run subsys=bgp-control-plane
[07:37:34] level=info msg="Reconciling peers for virtual router with local ASN 65001" component=manager.neighborReconciler subsys=bgp-control-plane
[07:37:34] level=info msg="Adding peer 172.16.100.3/32 65012 to local ASN 65001" component=manager.neighborReconciler subsys=bgp-control-plane
[07:37:34] level=info msg="Add a peer configuration" Key=172.16.100.3 Topic=Peer asn=65001 component=gobgp.BgpServerInstance subsys=bgp-control-plane
[07:37:34] level=info msg="Done reconciling peers for virtual router with local ASN 65001" component=manager.neighborReconcilersubsys=bgp-control-plane
[07:37:34] level=info msg="type:STATE  peer:{conf:{local_asn:65001  neighbor_address:\"172.16.100.2\"  peer_asn:65011}  state:{local_asn:65001 neighbor_address:\"172.16.100.2\"  peer_asn:65011  session_state:IDLE  router_id:\"<nil>\"}  transport:{local_address:\"<nil>\"}}"
[07:37:34] level=info msg="type:STATE  peer:{conf:{local_asn:65001  neighbor_address:\"172.16.100.2\"  peer_asn:65011}  state:{local_asn:65001 neighbor_address:\"172.16.100.2\"  peer_asn:65011  session_state:ACTIVE  router_id:\"<nil>\"}  transport:{local_address:\"<nil>\"}}"
[07:37:34] level=info msg="type:STATE  peer:{conf:{local_asn:65001  neighbor_address:\"172.16.100.3\"  peer_asn:65012}  state:{local_asn:65001 neighbor_address:\"172.16.100.3\"  peer_asn:65012  session_state:IDLE  router_id:\"<nil>\"}  transport:{local_address:\"<nil>\"}}"
[07:37:34] level=info msg="type:STATE  peer:{conf:{local_asn:65001  neighbor_address:\"172.16.100.3\"  peer_asn:65012}  state:{local_asn:65001 neighbor_address:\"172.16.100.3\"  peer_asn:65012  session_state:ACTIVE  router_id:\"<nil>\"}  transport:{local_address:\"<nil>\"}}"
[07:37:34] time="2023-05-24T07:33:46Z" level=debug msg="try to connect" Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGP subsys=basic
[07:37:34] time="2023-05-24T07:33:46Z" level=debug msg="state changed" Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGP new=BGP_FSM_OPENSENTold=BGP_FSM_ACTIVE reason=new-connection subsys=basic
[07:37:34] time="2023-05-24T07:33:46Z" level=info msg="GoBGP test instance: Peer Event: {65001 OPENSENT}"
[07:37:34] level=info msg="type:STATE  peer:{conf:{local_asn:65001  neighbor_address:\"172.16.100.2\"  peer_asn:65011}  state:{local_asn:65001 neighbor_address:\"172.16.100.2\"  peer_asn:65011  session_state:OPENSENT  router_id:\"<nil>\"}  transport:{local_address:\"172.16.100.1\" local_port:1790  remote_port:40939}}"
[07:37:34] level=info msg="type:STATE  peer:{conf:{local_asn:65001  neighbor_address:\"172.16.100.2\"  peer_asn:65011}  state:{local_asn:65001 neighbor_address:\"172.16.100.2\"  peer_asn:65011  session_state:OPENCONFIRM  router_id:\"172.16.100.2\"}  transport:{local_address:\"172.16.100.1\" local_port:1790  remote_port:40939}}"
[07:37:34] time="2023-05-24T07:33:46Z" level=debug msg="state changed" Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGPnew=BGP_FSM_OPENCONFIRM old=BGP_FSM_OPENSENT reason=open-msg-received subsys=basic
[07:37:34] time="2023-05-24T07:33:46Z" level=info msg="GoBGP test instance: Peer Event: {65001 OPENCONFIRM}"
[07:37:34] time="2023-05-24T07:33:46Z" level=info msg="Peer Up" Key=172.16.100.1 State=BGP_FSM_OPENCONFIRM Topic=Peer asn=65011 component=tests.BGPsubsys=basic
[07:37:34] time="2023-05-24T07:33:46Z" level=debug msg="state changed" Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGPnew=BGP_FSM_ESTABLISHED old=BGP_FSM_OPENCONFIRM reason=open-msg-negotiated subsys=basic
[07:37:34] time="2023-05-24T07:33:46Z" level=info msg="GoBGP test instance: Peer Event: {65001 ESTABLISHED}"
[07:37:34] time="2023-05-24T07:33:46Z" level=info msg="GoBGP test instance: Peer Event: {0 IDLE}"
[07:37:34] time="2023-05-24T07:33:46Z" level=info msg="GoBGP test instance: Peer Event: {65001 IDLE}"
[07:37:34] time="2023-05-24T07:33:46Z" level=info msg="GoBGP test instance: Peer Event: {65001 ACTIVE}"
[07:37:34] level=info msg="Peer Up" Key=172.16.100.2 State=BGP_FSM_OPENCONFIRM Topic=Peer asn=65001 component=gobgp.BgpServerInstancesubsys=bgp-control-plane
[07:37:34] level=info msg="type:STATE  peer:{conf:{local_asn:65001  neighbor_address:\"172.16.100.2\"  peer_asn:65011}  state:{local_asn:65001 neighbor_address:\"172.16.100.2\"  peer_asn:65011  session_state:ESTABLISHED  router_id:\"172.16.100.2\"}  transport:{local_address:\"172.16.100.1\" local_port:1790  remote_port:40939}}"
[07:37:34] time="2023-05-24T07:33:48Z" level=debug msg="try to connect" Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGP subsys=basic
[07:37:34] time="2023-05-24T07:33:48Z" level=debug msg="state changed" Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGP new=BGP_FSM_OPENSENTold=BGP_FSM_ACTIVE reason=new-connection subsys=basic
[07:37:34] time="2023-05-24T07:33:48Z" level=info msg="GoBGP test instance: Peer Event: {65001 OPENSENT}"
[07:37:34] level=info msg="type:STATE  peer:{conf:{local_asn:65001  neighbor_address:\"172.16.100.3\"  peer_asn:65012}  state:{local_asn:65001 neighbor_address:\"172.16.100.3\"  peer_asn:65012  session_state:OPENSENT  router_id:\"<nil>\"}  transport:{local_address:\"172.16.100.1\" local_port:1790  remote_port:56241}}"
[07:37:34] level=info msg="type:STATE  peer:{conf:{local_asn:65001  neighbor_address:\"172.16.100.3\"  peer_asn:65012}  state:{local_asn:65001 neighbor_address:\"172.16.100.3\"  peer_asn:65012  session_state:OPENCONFIRM  router_id:\"172.16.100.2\"}  transport:{local_address:\"172.16.100.1\" local_port:1790  remote_port:56241}}"
[07:37:34] time="2023-05-24T07:33:48Z" level=debug msg="state changed" Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGPnew=BGP_FSM_OPENCONFIRM old=BGP_FSM_OPENSENT reason=open-msg-received subsys=basic
[07:37:34] time="2023-05-24T07:33:48Z" level=info msg="GoBGP test instance: Peer Event: {65001 OPENCONFIRM}"
[07:37:34] time="2023-05-24T07:33:48Z" level=info msg="Peer Up" Key=172.16.100.1 State=BGP_FSM_OPENCONFIRM Topic=Peer asn=65011 component=tests.BGPsubsys=basic
[07:37:34] time="2023-05-24T07:33:48Z" level=debug msg="state changed" Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGPnew=BGP_FSM_ESTABLISHED old=BGP_FSM_OPENCONFIRM reason=open-msg-negotiated subsys=basic
[07:37:34] time="2023-05-24T07:33:48Z" level=info msg="GoBGP test instance: Peer Event: {65001 ESTABLISHED}"
[07:37:34] level=info msg="Peer Up" Key=172.16.100.3 State=BGP_FSM_OPENCONFIRM Topic=Peer asn=65001 component=gobgp.BgpServerInstancesubsys=bgp-control-plane
[07:37:34]     neighbor_test.go:117: 
[07:37:34]         	Error Trace:	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bgpv1/test/neighbor_test.go:117
[07:37:34]         	Error:      	elements differ
[07:37:34]         	            	
[07:37:34]         	            	extra elements in list A:
[07:37:34]         	            	([]interface {}) (len=1) {
[07:37:34]         	            	 (test.peeringState) {
[07:37:34]         	            	  peerASN: (uint32) 65012,
[07:37:34]         	            	  peerAddr: (string) (len=12) "172.16.100.3",
[07:37:34]         	            	  peerSession: (string) (len=11) "established"
[07:37:34]         	            	 }
[07:37:34]         	            	}
[07:37:34]         	            	
[07:37:34]         	            	
[07:37:34]         	            	extra elements in list B:
[07:37:34]         	            	([]interface {}) (len=1) {
[07:37:34]         	            	 (test.peeringState) {
[07:37:34]         	            	  peerASN: (uint32) 65012,
[07:37:34]         	            	  peerAddr: (string) (len=12) "172.16.100.3",
[07:37:34]         	            	  peerSession: (string) (len=12) "open_confirm"
[07:37:34]         	            	 }
[07:37:34]         	            	}
[07:37:34]         	            	
[07:37:34]         	            	
[07:37:34]         	            	listA:
[07:37:34]         	            	([]test.peeringState) (len=2) {
[07:37:34]         	            	 (test.peeringState) {
[07:37:34]         	            	  peerASN: (uint32) 65011,
[07:37:34]         	            	  peerAddr: (string) (len=12) "172.16.100.2",
[07:37:34]         	            	  peerSession: (string) (len=11) "established"
[07:37:34]         	            	 },
[07:37:34]         	            	 (test.peeringState) {
[07:37:34]         	            	  peerASN: (uint32) 65012,
[07:37:34]         	            	  peerAddr: (string) (len=12) "172.16.100.3",
[07:37:34]         	            	  peerSession: (string) (len=11) "established"
[07:37:34]         	            	 }
[07:37:34]         	            	}
[07:37:34]         	            	
[07:37:34]         	            	
[07:37:34]         	            	listB:
[07:37:34]         	            	([]test.peeringState) (len=2) {
[07:37:34]         	            	 (test.peeringState) {
[07:37:34]         	            	  peerASN: (uint32) 65011,
[07:37:34]         	            	  peerAddr: (string) (len=12) "172.16.100.2",
[07:37:34]         	            	  peerSession: (string) (len=11) "established"
[07:37:34]         	            	 },
[07:37:34]         	            	 (test.peeringState) {
[07:37:34]         	            	  peerASN: (uint32) 65012,
[07:37:34]         	            	  peerAddr: (string) (len=12) "172.16.100.3",
[07:37:34]         	            	  peerSession: (string) (len=12) "open_confirm"
[07:37:34]         	            	 }
[07:37:34]         	            	}
[07:37:34]         	Test:       	Test_NeighborAddDel/add_two_neighbors
[07:37:34]         	Messages:   	add two neighbors
[07:37:34] level=info msg="type:STATE  peer:{conf:{local_asn:65001  neighbor_address:\"172.16.100.3\"  peer_asn:65012}  state:{local_asn:65001 neighbor_address:\"172.16.100.3\"  peer_asn:65012  session_state:ESTABLISHED  router_id:\"172.16.100.2\"}  transport:{local_address:\"172.16.100.1\" local_port:1790  remote_port:56241}}"
```

### PR DESCRIPTION
**Pull request by rastislavs**
Fixes flakes due to potential timing issue in the `Test_NeighborAddDel` test.

In the test we wait for peers to transition into their expected state. However, remote peer's session state does not have to immediately match our session's state (e.g. peer may be already in Established but we still in OpenConfirm until we receive a Keepalive from the peer), so we should retry the check with a timeout if our state of the session does not immediately match.

Fixes: #25637

### CODE REVIEW

File: pkg/bgpv1/test/neighbor_test.go [hunk 1]
```diff
@@ -100,21 +102,32 @@ func Test_NeighborAddDel(t *testing.T) {
...
-				})
+				var runningState []peeringState
+				for _, peer := range peers {
+					runningState = append(runningState, peeringState{
+						peerASN:     uint32(peer.PeerAsn),
+						peerAddr:    peer.PeerAddress,
+						peerSession: peer.SessionState,
+					})
+				}
+				return assert.ElementsMatch(t, step.expectedPeerStates, runningState, step.description)
```
Comment (harsimran-pabla) [inline pkg/bgpv1/test/neighbor_test.go:123]: Any specific reason to use assert instead of require ?

**Review by harsimran-pabla — APPROVED** [review]
Thank you!

File: pkg/bgpv1/test/neighbor_test.go (hunk 1 above)
Comment (rastislavs) [reply pkg/bgpv1/test/neighbor_test.go:123]: EDIT (reworded): require just fails the test, assert returns bool

**Comment by ldelossa** [comment]
Just out of curiosity, if we suspect a flake, was this test ran over multiple trials to see if the flake was removed?

**Comment by rastislavs** [comment]
> Just out of curiosity, if we suspect a flake, was this test ran over multiple trials to see if the flake was removed?

This is a bit tricky, as I could not reproduce the flake when running it locally (not before nor after fix). Somehow it is reproducible only in CI - we can re-run the CI test multiple times before merging.

**Review by ldelossa — APPROVED** [review]

**Comment by rastislavs** [comment]
The `test-1.26-net-next` job is failing due to https://github.com/cilium/cilium/issues/25605
All other are green.

It seems that the `test-runtime` job that was flaking due to this issue has been pretty successful in the past runs, so not sure if re-running it once or twice would even prove anything. Maybe we should just merge this and observe it over a larger time period.

**Comment by harsimran-pabla** [comment]
> It seems that the test-runtime job that was flaking due to this issue has been pretty successful in the past runs, so not sure if re-running it once or twice would even prove anything. Maybe we should just merge this and observe it over a larger time period.

Locally I was able to reproduce same failure few time running like this 
`PRIVILEGED_TESTS=1 go test -exec 'sudo -E' ./... -count 10 -run Test_NeighborAddDel` from bgpv1/test directory. 

You can try to validate if this change fixes it.

**Comment by rastislavs** [comment]
> Locally I was able to reproduce same failure few time running like this `PRIVILEGED_TESTS=1 go test -exec 'sudo -E' ./... -count 10 -run Test_NeighborAddDel` from bgpv1/test directory.
> 
> You can try to validate if this change fixes it.

@harsimran-pabla Thanks for the tip, I was able to reproduce the issue (without the fix) with some more iterations and it seems that the fix indeed helped as I could not reproduce it with 10x more iterations with the fix.

**Comment by rastislavs** [comment]
As I was able to verify the fix locally and the `test-1.26-net-next` job is failing due to https://github.com/cilium/cilium/issues/25605, I think this is good to go, marking as ready-to-merge.

**Pull request merged**